File: backend/audio_sequence_integration.py
"""
StoryCore-Engine Audio-Sequence Integration Module

This module integrates the multi-track audio system with the sequence generation system.
It automatically creates audio profiles based on sequence context and enables
synchronized audio generation for generated sequences.

Integration Features:
- Auto-profile generation from sequence context
- Shot-level audio attachment
- Timeline synchronization
- Cross-module data flow

Requirements: Q1 2026 - Audio/Sequence Integration
"""

import logging
from typing import Any, Dict, List, Optional
from datetime import datetime
from enum import Enum
from dataclasses import dataclass, field

from backend.music_profile_builder import MusicProfileBuilder
from backend.sfx_profile_builder import SFXProfileBuilder
from backend.voice_profile_builder import VoiceProfileBuilder
from backend.prompt_composer import PromptComposer

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class SequenceAudioType(str, Enum):
    """Audio types for sequence integration"""
    MUSIC = "music"
    SFX = "sfx"
    VOICE = "voice"
    ALL = "all"


class IntegrationStatus(str, Enum):
    """Status of audio-sequence integration"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class SequenceAudioContext:
    """
    Audio context derived from sequence/shot data.
    This bridges the gap between sequence metadata and audio generation.
    """
    # Project context
    project_id: str
    project_type: str = ""
    location: str = ""
    themes: List[str] = field(default_factory=list)
    visual_style: str = ""
    emotional_intensity: str = "medium"
    
    # Shot context (per-shot audio)
    shot_id: Optional[str] = None
    shot_type: str = ""
    shot_description: str = ""
    shot_duration: float = 0.0
    
    # Action context
    action_type: str = ""
    action_intensity: str = "medium"
    visual_rhythm: str = "normal"
    
    # Beat context
    beat_type: str = ""
    beat_importance: str = "medium"
    
    # Timing
    start_time: float = 0.0
    end_time: float = 0.0
    
    # Options
    auto_mix: bool = True
    ducking: bool = True


@dataclass
class IntegratedAudioResult:
    """Result of audio-sequence integration"""
    success: bool
    status: IntegrationStatus
    music_profile: Optional[Dict[str, Any]] = None
    sfx_profile: Optional[Dict[str, Any]] = None
    voice_profile: Optional[Dict[str, Any]] = None
    music_prompts: List[Dict[str, str]] = field(default_factory=list)
    sfx_prompts: List[Dict[str, str]] = field(default_factory=list)
    voice_prompts: List[Dict[str, str]] = field(default_factory=list)
    mix_configuration: Optional[Dict[str, Any]] = None
    errors: List[str] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)
    generated_at: str = ""


class AudioSequenceIntegrator:
    """
    Main integrator class for audio-sequence functionality.
    
    This class bridges the sequence/shot generation system with the
    multi-track audio generation system.
    """
    
    def __init__(self):
        """Initialize the integrator"""
        self.prompt_composer = PromptComposer()
        logger.info("Audio-Sequence Integrator initialized")
    
    def extract_context_from_sequence(
        self,
        sequence: Dict[str, Any],
        project_context: Optional[Dict[str, Any]] = None
    ) -> List[SequenceAudioContext]:
        """
        Extract audio contexts from a generated sequence.
        
        Args:
            sequence: Generated sequence data
            project_context: Optional project-level context
        
        Returns:
            List of SequenceAudioContext for each shot
        """
        contexts = []
        project_id = sequence.get('project_id', '')
        shots = sequence.get('shots', [])
        
        # Extract project context
        proj_type = project_context.get('project_type', '') if project_context else ''
        location = project_context.get('location', '') if project_context else ''
        themes = project_context.get('themes', []) if project_context else []
        style = project_context.get('visual_style', '') if project_context else ''
        intensity = project_context.get('emotional_intensity', 'medium') if project_context else 'medium'
        
        for shot in shots:
            context = SequenceAudioContext(
                project_id=project_id,
                project_type=proj_type,
                location=location,
                themes=themes.copy(),
                visual_style=style,
                emotional_intensity=intensity,
                shot_id=shot.get('id'),
                shot_type=shot.get('shot_type', 'action'),
                shot_description=shot.get('prompt', ''),
                shot_duration=shot.get('duration_seconds', 5.0)
            )
            contexts.append(context)
        
        logger.info(f"Extracted {len(contexts)} audio contexts from sequence")
        return contexts
    
    def create_music_profile_from_context(
        self,
        context: SequenceAudioContext
    ) -> Dict[str, Any]:
        """
        Create a music profile from sequence context.
        
        Args:
            context: SequenceAudioContext
        
        Returns:
            Music profile as dictionary
        """
        builder = MusicProfileBuilder(context.project_id)
        
        if context.project_type:
            builder.set_project_type(context.project_type)
        if context.location:
            builder.set_location(context.location)
        for theme in context.themes:
            builder.add_theme(theme)
        if context.visual_style:
            builder.set_visual_style(context.visual_style)
        if context.emotional_intensity:
            builder.set_emotional_intensity(context.emotional_intensity)
        
        if context.action_type:
            builder.set_action(
                context.action_type,
                context.action_intensity,
                context.visual_rhythm
            )
        
        if context.shot_duration:
            builder.set_duration(context.shot_duration)
        
        profile = builder.build()
        return profile.to_dict() if hasattr(profile, 'to_dict') else profile.__dict__
    
    def create_sfx_profile_from_context(
        self,
        context: SequenceAudioContext
    ) -> Dict[str, Any]:
        """
        Create an SFX profile from sequence context.
        
        Args:
            context: SequenceAudioContext
        
        Returns:
            SFX profile as dictionary
        """
        builder = SFXProfileBuilder(context.project_id)
        
        action_mapping = {
            'action': 'fight',
            'dialogue': 'footsteps',
            'ambient': 'wind',
            'transition': 'whoosh'
        }
        sfx_action = action_mapping.get(context.shot_type, 'impact')
        
        intensity_mapping = {
            'low': 'low',
            'medium': 'medium',
            'high': 'high'
        }
        intensity = intensity_mapping.get(context.action_intensity, 'medium')
        
        builder.set_action_type(sfx_action)
        builder.set_intensity(intensity)
        if context.location:
            builder.set_environment(context.location)
        builder.set_duration(context.shot_duration)
        
        if context.shot_type in ['underwater', 'submarine']:
            builder.enable_muffling()
        
        profile = builder.build()
        return profile.to_dict() if hasattr(profile, 'to_dict') else profile.__dict__
    
    def create_voice_profile_from_context(
        self,
        context: SequenceAudioContext,
        text_content: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Create a voice profile from sequence context.
        
        Args:
            context: SequenceAudioContext
            text_content: Optional text to speak
        
        Returns:
            Voice profile as dictionary
        """
        builder = VoiceProfileBuilder(context.project_id)
        
        voice_mapping = {
            'dialogue': 'raw',
            'narration': 'whisper',
            'action': 'raw'
        }
        voice_type = voice_mapping.get(context.shot_type, 'raw')
        
        builder.set_voice_type(voice_type)
        if text_content:
            builder.set_text(text_content)
        builder.set_language('fr')
        if context.shot_duration:
            builder.set_duration(context.shot_duration)
        
        profile = builder.build()
        return profile.to_dict() if hasattr(profile, 'to_dict') else profile.__dict__
    
    def generate_audio_for_sequence(
        self,
        sequence: Dict[str, Any],
        project_context: Optional[Dict[str, Any]] = None,
        audio_types: List[SequenceAudioType] = None
    ) -> IntegratedAudioResult:
        """
        Generate all audio for a complete sequence.
        
        Args:
            sequence: Generated sequence data
            project_context: Optional project-level context
            audio_types: Types of audio to generate
        
        Returns:
            IntegratedAudioResult with all generated profiles and prompts
        """
        if audio_types is None:
            audio_types = [SequenceAudioType.ALL]
        
        result = IntegratedAudioResult(
            success=True,
            status=IntegrationStatus.COMPLETED,
            generated_at=datetime.utcnow().isoformat()
        )
        
        contexts = self.extract_context_from_sequence(sequence, project_context)
        
        if SequenceAudioType.MUSIC in audio_types or SequenceAudioType.ALL in audio_types:
            for context in contexts:
                music_profile = self.create_music_profile_from_context(context)
                result.music_profile = music_profile
                
                try:
                    builder = MusicProfileBuilder(context.project_id)
                    if context.project_type:
                        builder.set_project_type(context.project_type)
                    if context.location:
                        builder.set_location(context.location)
                    for theme in context.themes:
                        builder.add_theme(theme)
                    if context.visual_style:
                        builder.set_visual_style(context.visual_style)
                    if context.emotional_intensity:
                        builder.set_emotional_intensity(context.emotional_intensity)
                    if context.action_type:
                        builder.set_action(context.action_type, context.action_intensity, context.visual_rhythm)
                    builder.set_duration(context.shot_duration)
                    
                    profile = builder.build()
                    prompts = self.prompt_composer.compose_all_music_prompts(profile)
                    result.music_prompts.extend([
                        {"track_name": p.track_name, "track_type": p.track_type, "prompt": p.prompt, "shot_id": context.shot_id}
                        for p in prompts
                    ])
                except Exception as e:
                    result.errors.append(f"Music prompt generation failed: {str(e)}")
        
        if SequenceAudioType.SFX in audio_types or SequenceAudioType.ALL in audio_types:
            for context in contexts:
                sfx_profile = self.create_sfx_profile_from_context(context)
                result.sfx_profile = sfx_profile
                
                try:
                    builder = SFXProfileBuilder(context.project_id)
                    if context.action_type:
                        builder.set_action_type(context.action_type)
                    builder.set_intensity(context.action_intensity)
                    if context.location:
                        builder.set_environment(context.location)
                    builder.set_duration(context.shot_duration)
                    
                    profile = builder.build()
                    prompts = self.prompt_composer.compose_all_sfx_prompts(profile)
                    result.sfx_prompts.extend([
                        {"track_name": p.track_name, "track_type": p.track_type, "prompt": p.prompt, "shot_id": context.shot_id}
                        for p in prompts
                    ])
                except Exception as e:
                    result.errors.append(f"SFX prompt generation failed: {str(e)}")
        
        if SequenceAudioType.VOICE in audio_types or SequenceAudioType.ALL in audio_types:
            for context in contexts:
                voice_profile = self.create_voice_profile_from_context(context)
                result.voice_profile = voice_profile
                
                try:
                    builder = VoiceProfileBuilder(context.project_id)
                    builder.set_voice_type('raw')
                    if context.shot_description:
                        sample_text = context.shot_description[:200]
                        builder.set_text(sample_text)
                    builder.set_language('fr')
                    builder.set_duration(context.shot_duration)
                    
                    profile = builder.build()
                    prompt = self.prompt_composer.compose_voice_prompt(profile)
                    result.voice_prompts.append({
                        "track_name": prompt.track_name,
                        "track_type": prompt.track_type,
                        "prompt": prompt.prompt,
                        "shot_id": context.shot_id
                    })
                except Exception as e:
                    result.errors.append(f"Voice prompt generation failed: {str(e)}")
        
        result.success = len(result.errors) == 0
        result.status = IntegrationStatus.COMPLETED if result.success else IntegrationStatus.FAILED
        
        logger.info(f"Audio-sequence integration: {len(result.music_prompts)} music, {len(result.sfx_prompts)} SFX, {len(result.voice_prompts)} voice")
        
        return result
    
    def attach_audio_to_shot(
        self,
        shot: Dict[str, Any],
        audio_result: IntegratedAudioResult,
        audio_type: SequenceAudioType = SequenceAudioType.ALL
    ) -> Dict[str, Any]:
        """
        Attach generated audio to a shot for timeline integration.
        
        Args:
            shot: Shot data to attach audio to
            audio_result: IntegratedAudioResult from generation
            audio_type: Type of audio to attach
        
        Returns:
            Shot with audio attached
        """
        shot_id = shot.get('id')
        
        music_tracks = [p for p in audio_result.music_prompts if p.get('shot_id') == shot_id]
        sfx_tracks = [p for p in audio_result.sfx_prompts if p.get('shot_id') == shot_id]
        voice_tracks = [p for p in audio_result.voice_prompts if p.get('shot_id') == shot_id]
        
        audio_tracks = []
        
        if audio_type in [SequenceAudioType.MUSIC, SequenceAudioType.ALL]:
            for track in music_tracks:
                audio_tracks.append({
                    "id": f"{shot_id}_music_{track['track_name'].lower().replace(' ', '_')}",
                    "name": track['track_name'],
                    "type": "music",
                    "prompt": track['prompt'],
                    "start_time": 0,
                    "duration": shot.get('duration_seconds', 5.0),
                    "volume": -6.0,
                    "muted": False,
                    "regeneratable": True
                })
        
        if audio_type in [SequenceAudioType.SFX, SequenceAudioType.ALL]:
            for track in sfx_tracks:
                audio_tracks.append({
                    "id": f"{shot_id}_sfx_{track['track_name'].lower().replace(' ', '_')}",
                    "name": track['track_name'],
                    "type": "sfx",
                    "prompt": track['prompt'],
                    "start_time": 0,
                    "duration": shot.get('duration_seconds', 5.0),
                    "volume": 0.0,
                    "muted": False,
                    "regeneratable": True
                })
        
        if audio_type in [SequenceAudioType.VOICE, SequenceAudioType.ALL]:
            for track in voice_tracks:
                audio_tracks.append({
                    "id": f"{shot_id}_voice_{track['track_name'].lower().replace(' ', '_')}",
                    "name": track['track_name'],
                    "type": "voice",
                    "prompt": track['prompt'],
                    "start_time": 0,
                    "duration": shot.get('duration_seconds', 5.0),
                    "volume": 0.0,
                    "muted": False,
                    "regeneratable": True
                })
        
        shot['audio_tracks'] = audio_tracks
        shot['audio_generated'] = len(audio_tracks) > 0
        
        return shot
    
    def generate_timeline_sync_data(
        self,
        sequence: Dict[str, Any],
        audio_result: IntegratedAudioResult
    ) -> Dict[str, Any]:
        """
        Generate timeline synchronization data for the video editor.
        
        Args:
            sequence: Original sequence
            audio_result: Generated audio result
        
        Returns:
            Timeline sync data with all tracks and timing
        """
        shots = sequence.get('shots', [])
        timeline_tracks = []
        
        track_types = [
            ("dialogue", "Dialogue"),
            ("sfx_action", "SFX - Action"),
            ("sfx_environment", "SFX - Environment"),
            ("music_base", "Music - Base"),
            ("music_melody", "Music - Melody"),
            ("music_percussion", "Music - Percussion"),
            ("music_bass", "Music - Bass"),
            ("music_fx", "Music - FX"),
            ("ambient", "Ambient")
        ]
        
        for track_id, track_name in track_types:
            timeline_tracks.append({
                "id": f"timeline_{track_id}",
                "name": track_name,
                "type": track_id.split("_")[0] if "_" in track_id else track_id,
                "clips": [],
                "muted": False,
                "volume": 0.0,
                "height": 40
            })
        
        current_time = 0.0
        for shot in shots:
            shot_duration = shot.get('duration_seconds', 5.0)
            shot_id = shot.get('id', '')
            audio_tracks = shot.get('audio_tracks', [])
            
            for audio_track in audio_tracks:
                track_type = audio_track.get('type', 'music')
                track_name = audio_track.get('name', '')
                
                if track_type == "voice":
                    timeline_track_id = "timeline_dialogue"
                elif track_type == "sfx":
                    if "action" in track_name.lower():
                        timeline_track_id = "timeline_sfx_action"
                    else:
                        timeline_track_id = "timeline_sfx_environment"
                else:
                    if "base" in track_name.lower() or "fondation" in track_name.lower():
                        timeline_track_id = "timeline_music_base"
                    elif "melody" in track_name.lower() or "mélodie" in track_name.lower():
                        timeline_track_id = "timeline_music_melody"
                    elif "percussion" in track_name.lower():
                        timeline_track_id = "timeline_music_percussion"
                    elif "bass" in track_name.lower() or "basse" in track_name.lower():
                        timeline_track_id = "timeline_music_bass"
                    elif "fx" in track_name.lower():
                        timeline_track_id = "timeline_music_fx"
                    else:
                        timeline_track_id = "timeline_music_base"
                
                for track in timeline_tracks:
                    if track["id"] == timeline_track_id:
                        track["clips"].append({
                            "id": audio_track["id"],
                            "name": track_name,
                            "start_time": current_time,
                            "duration": shot_duration,
                            "audio_track_id": audio_track["id"],
                            "prompt": audio_track.get('prompt', ''),
                            "regeneratable": audio_track.get('regeneratable', True)
                        })
                        break
            
            current_time += shot_duration
        
        return {
            "sequence_id": sequence.get('id', ''),
            "project_id": sequence.get('project_id', ''),
            "total_duration": current_time,
            "tracks": timeline_tracks,
            "generated_at": datetime.utcnow().isoformat(),
            "audio_prompts": {
                "music": audio_result.music_prompts,
                "sfx": audio_result.sfx_prompts,
                "voice": audio_result.voice_prompts
            }
        }


def create_audio_sequence_integrator() -> AudioSequenceIntegrator:
    """Create a new AudioSequenceIntegrator instance."""
    return AudioSequenceIntegrator()

File: creative-studio-ui/src/services/audioSequenceIntegration.ts
/**
 * StoryCore-Engine Audio-Sequence Integration Service
 */

import type {
  MusicProfile,
  SFXProfile,
  VoiceProfile,
  MixConfiguration
} from '../types/audioMultitrack';
import { audioMultitrackService } from './audioMultitrack';

export type IntegrationAudioType = 'music' | 'sfx' | 'voice' | 'all';
export type IntegrationStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface SequenceAudioContext {
  projectId: string;
  projectType?: string;
  location?: string;
  themes: string[];
  visualStyle?: string;
  emotionalIntensity: string;
  shotId?: string;
  shotType: string;
  shotDescription?: string;
  shotDuration: number;
  actionType?: string;
  actionIntensity?: string;
  visualRhythm?: string;
  autoMix: boolean;
  ducking: boolean;
}

export interface IntegratedAudioResult {
  success: boolean;
  status: IntegrationStatus;
  musicPrompts: TrackPrompt[];
  sfxPrompts: TrackPrompt[];
  voicePrompts: TrackPrompt[];
  errors: string[];
  generatedAt: string;
}

export interface TrackPrompt {
  trackName: string;
  trackType: string;
  prompt: string;
  shotId?: string;
}

export interface TimelineTrack {
  id: string;
  name: string;
  type: string;
  clips: TimelineClip[];
  muted: boolean;
  volume: number;
  height: number;
}

export interface TimelineClip {
  id: string;
  name: string;
  startTime: number;
  duration: number;
  audioTrackId: string;
  prompt: string;
  regeneratable: boolean;
}

export interface TimelineSyncData {
  sequenceId: string;
  projectId: string;
  totalDuration: number;
  tracks: TimelineTrack[];
  generatedAt: string;
  audioPrompts: {
    music: TrackPrompt[];
    sfx: TrackPrompt[];
    voice: TrackPrompt[];
  };
}

export class AudioSequenceIntegrationService {
  /**
   * Extract audio contexts from a generated sequence
   */
  extractContextsFromSequence(
    sequence: unknown,
    projectContext?: Record<string, unknown>
  ): SequenceAudioContext[] {
    const contexts: SequenceAudioContext[] = [];
    const projectId = sequence.project_id || '';
    const shots = sequence.shots || [];
    const projType = projectContext?.project_type || '';
    const location = projectContext?.location || '';
    const themes = projectContext?.themes || [];
    const style = projectContext?.visual_style || '';
    const intensity = projectContext?.emotional_intensity || 'medium';

    for (const shot of shots) {
      contexts.push({
        projectId,
        projectType: projType,
        location,
        themes: [...themes],
        visualStyle: style,
        emotionalIntensity: intensity,
        shotId: shot.id,
        shotType: shot.shot_type || 'action',
        shotDescription: shot.prompt || '',
        shotDuration: shot.duration_seconds || 5.0,
        autoMix: true,
        ducking: true
      });
    }
    return contexts;
  }

  /**
   * Generate all audio for a complete sequence
   */
  async generateAudioForSequence(
    sequence: unknown,
    projectContext?: Record<string, unknown>,
    audioTypes: IntegrationAudioType[] = ['all']
  ): Promise<IntegratedAudioResult> {
    const result: IntegratedAudioResult = {
      success: true,
      status: 'completed',
      musicPrompts: [],
      sfxPrompts: [],
      voicePrompts: [],
      errors: [],
      generatedAt: new Date().toISOString()
    };

    const contexts = this.extractContextsFromSequence(sequence, projectContext);
    const shouldGenerateMusic = audioTypes.includes('all') || audioTypes.includes('music');
    const shouldGenerateSFX = audioTypes.includes('all') || audioTypes.includes('sfx');
    const shouldGenerateVoice = audioTypes.includes('all') || audioTypes.includes('voice');

    if (shouldGenerateMusic) {
      for (const context of contexts) {
        result.musicPrompts.push(
          { trackName: 'Base', trackType: 'base', prompt: `Music base for ${context.shotType}`, shotId: context.shotId },
          { trackName: 'Melody', trackType: 'melody', prompt: `Music melody for ${context.shotType}`, shotId: context.shotId },
          { trackName: 'Percussion', trackType: 'percussion', prompt: `Percussion for ${context.shotType}`, shotId: context.shotId },
          { trackName: 'Bass', trackType: 'bass', prompt: `Bass for ${context.shotType}`, shotId: context.shotId },
          { trackName: 'FX', trackType: 'fx', prompt: `Music FX for ${context.shotType}`, shotId: context.shotId },
          { trackName: 'Drones', trackType: 'drones', prompt: `Drones for ${context.shotType}`, shotId: context.shotId }
        );
      }
    }

    if (shouldGenerateSFX) {
      for (const context of contexts) {
        result.sfxPrompts.push(
          { trackName: 'Action', trackType: 'action', prompt: `SFX for ${context.shotType} action`, shotId: context.shotId },
          { trackName: 'Environment', trackType: 'environment', prompt: `Environment sounds for ${context.location || 'scene'}`, shotId: context.shotId },
          { trackName: 'Stylized', trackType: 'stylized', prompt: `Stylized SFX for ${context.shotType}`, shotId: context.shotId },
          { trackName: 'Bullet Time', trackType: 'bullet_time', prompt: `Bullet time SFX for ${context.shotType}`, shotId: context.shotId }
        );
      }
    }

    if (shouldGenerateVoice) {
      for (const context of contexts) {
        result.voicePrompts.push({
          trackName: 'Voice',
          trackType: context.shotType === 'narration' ? 'whisper' : 'raw',
          prompt: context.shotDescription || '',
          shotId: context.shotId
        });
      }
    }

    result.status = result.success ? 'completed' : 'failed';
    return result;
  }

  /**
   * Attach generated audio to a shot
   */
  attachAudioToShot(
    shot: unknown,
    audioResult: IntegratedAudioResult,
    audioType: IntegrationAudioType = 'all'
  ): unknown {
    const shotId = shot.id;
    const musicTracks = audioResult.musicPrompts.filter(p => p.shotId === shotId);
    const sfxTracks = audioResult.sfxPrompts.filter(p => p.shotId === shotId);
    const voiceTracks = audioResult.voicePrompts.filter(p => p.shotId === shotId);
    const audioTracks: unknown[] = [];
    const duration = shot.duration_seconds || 5.0;

    if (audioType === 'all' || audioType === 'music') {
      for (const track of musicTracks) {
        audioTracks.push({
          id: `${shotId}_music_${track.trackName.toLowerCase().replace(' ', '_')}`,
          name: track.trackName,
          type: 'music',
          prompt: track.prompt,
          startTime: 0,
          duration,
          volume: -6.0,
          muted: false,
          regeneratable: true
        });
      }
    }

    if (audioType === 'all' || audioType === 'sfx') {
      for (const track of sfxTracks) {
        audioTracks.push({
          id: `${shotId}_sfx_${track.trackName.toLowerCase().replace(' ', '_')}`,
          name: track.trackName,
          type: 'sfx',
          prompt: track.prompt,
          startTime: 0,
          duration,
          volume: 0.0,
          muted: false,
          regeneratable: true
        });
      }
    }

    if (audioType === 'all' || audioType === 'voice') {
      for (const track of voiceTracks) {
        audioTracks.push({
          id: `${shotId}_voice_${track.trackName.toLowerCase().replace(' ', '_')}`,
          name: track.trackName,
          type: 'voice',
          prompt: track.prompt,
          startTime: 0,
          duration,
          volume: 0.0,
          muted: false,
          regeneratable: true
        });
      }
    }

    return { ...shot, audio_tracks: audioTracks, audio_generated: audioTracks.length > 0 };
  }

  /**
   * Generate timeline synchronization data
   */
  generateTimelineSyncData(sequence: unknown, audioResult: IntegratedAudioResult): TimelineSyncData {
    const shots = sequence.shots || [];
    const timelineTracks: TimelineTrack[] = [
      { id: 'timeline_dialogue', name: 'Dialogue', type: 'voice', clips: [], muted: false, volume: 0, height: 40 },
      { id: 'timeline_sfx_action', name: 'SFX - Action', type: 'sfx', clips: [], muted: false, volume: 0, height: 40 },
      { id: 'timeline_sfx_environment', name: 'SFX - Environment', type: 'sfx', clips: [], muted: false, volume: 0, height: 40 },
      { id: 'timeline_music_base', name: 'Music - Base', type: 'music', clips: [], muted: false, volume: 0, height: 40 },
      { id: 'timeline_music_melody', name: 'Music - Melody', type: 'music', clips: [], muted: false, volume: 0, height: 40 },
      { id: 'timeline_music_percussion', name: 'Music - Percussion', type: 'music', clips: [], muted: false, volume: 0, height: 40 },
      { id: 'timeline_music_bass', name: 'Music - Bass', type: 'music', clips: [], muted: false, volume: 0, height: 40 },
      { id: 'timeline_music_fx', name: 'Music - FX', type: 'music', clips: [], muted: false, volume: 0, height: 40 }
    ];

    let currentTime = 0;
    for (const shot of shots) {
      const shotDuration = shot.duration_seconds || 5.0;
      const shotId = shot.id || '';
      const audioTracks = shot.audio_tracks || [];

      for (const audioTrack of audioTracks) {
        const trackType = audioTrack.type || 'music';
        const trackName = audioTrack.name || '';
        let timelineTrackId = '';

        if (trackType === 'voice') {
          timelineTrackId = 'timeline_dialogue';
        } else if (trackType === 'sfx') {
          timelineTrackId = trackName.toLowerCase().includes('action') ? 'timeline_sfx_action' : 'timeline_sfx_environment';
        } else {
          const nameLower = trackName.toLowerCase();
          if (nameLower.includes('base')) timelineTrackId = 'timeline_music_base';
          else if (nameLower.includes('melody')) timelineTrackId = 'timeline_music_melody';
          else if (nameLower.includes('percussion')) timelineTrackId = 'timeline_music_percussion';
          else if (nameLower.includes('bass')) timelineTrackId = 'timeline_music_bass';
          else if (nameLower.includes('fx')) timelineTrackId = 'timeline_music_fx';
          else timelineTrackId = 'timeline_music_base';
        }

        for (const track of timelineTracks) {
          if (track.id === timelineTrackId) {
            track.clips.push({
              id: audioTrack.id,
              name: trackName,
              startTime: currentTime,
              duration: shotDuration,
              audioTrackId: audioTrack.id,
              prompt: audioTrack.prompt || '',
              regeneratable: audioTrack.regeneratable !== false
            });
            break;
          }
        }
      }
      currentTime += shotDuration;
    }

    return {
      sequenceId: sequence.id || '',
      projectId: sequence.project_id || '',
      totalDuration: currentTime,
      tracks: timelineTracks,
      generatedAt: new Date().toISOString(),
      audioPrompts: { music: audioResult.musicPrompts, sfx: audioResult.sfxPrompts, voice: audioResult.voicePrompts }
    };
  }

  /**
   * Apply audio to entire sequence
   */
  async applyAudioToSequence(
    sequence: unknown,
    projectContext?: Record<string, unknown>,
    audioTypes: IntegrationAudioType[] = ['all']
  ): Promise<{ sequence: unknown; timeline: TimelineSyncData }> {
    const audioResult = await this.generateAudioForSequence(sequence, projectContext, audioTypes);
    const type = audioTypes.includes('all') ? 'all' : audioTypes[0];
    const updatedShots = (sequence.shots || []).map((shot: unknown) => this.attachAudioToShot(shot, audioResult, type));
    const updatedSequence = { ...sequence, shots: updatedShots, audio_generated: audioResult.success, audio_generation_result: audioResult };
    const timeline = this.generateTimelineSyncData(updatedSequence, audioResult);
    return { sequence: updatedSequence, timeline };
  }
}

export const audioSequenceIntegrationService = new AudioSequenceIntegrationService();

export function getDefaultAudioForShotType(shotType: string): { musicIntensity: string; sfxIntensity: string; enableVoice: boolean } {
  const mapping: Record<string, { musicIntensity: string; sfxIntensity: string; enableVoice: boolean }> = {
    'action': { musicIntensity: 'high', sfxIntensity: 'high', enableVoice: false },
    'dialogue': { musicIntensity: 'low', sfxIntensity: 'low', enableVoice: true },
    'ambient': { musicIntensity: 'medium', sfxIntensity: 'medium', enableVoice: false },
    'transition': { musicIntensity: 'low', sfxIntensity: 'low', enableVoice: false },
    'narration': { musicIntensity: 'low', sfxIntensity: 'low', enableVoice: true }
  };
  return mapping[shotType] || { musicIntensity: 'medium', sfxIntensity: 'medium', enableVoice: false };
}

export function estimateAudioGenerationTime(shotCount: number, audioTypes: IntegrationAudioType[]): number {
  return 30 + (shotCount * 10 * audioTypes.length);
}




